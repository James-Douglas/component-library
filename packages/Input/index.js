import Input from './Input.component';
import CurrencyInput from './Currency/CurrencyInput.component';
import DateInput from './Date/DateInput.component';
import ExpressiveInput from './Expressive/ExpressiveInput.component';

export {
  Input, DateInput, CurrencyInput, ExpressiveInput,
};
