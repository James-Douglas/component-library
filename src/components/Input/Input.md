# Input component<br/>
## Usage
~~~js
<Input 
    id="input-id"
    type="text"
    label="label"
    tooltip={
        title: 'Tooltip heading',
        body: 'Tooltip body',
    }
    placeholder="placeholder string"
    bordered={true}
    disabled={false}
    required={false}
    invalid={false}
    prefillValue=""
    prefixContent=""
    suffixContent=""
    listId="" //TODO
    autocomplete="off"
    handleChange={() => {}}
/>
~~~

## Props
The Input component can be extended by the combo component //TODO, and accepts the following props:

1. `id` *(required) string*
    * Unique id for the component. Required.
2. `type` *(optional) string*
    * Type of input (such as number, text, tel etc). Defaults to 'text'.
3. `label` *(optional) string*
    * Label for the input, provided by the fieldset component. Defaults to an empty string.
4. `tooltip` *(optional) object*
    * Tooltip for the fieldset, provided by the fieldset component. Accepts title and body properties. Defaults to an empty obj.
5. `placeholder` *(optional) string*
    * The placeholder text for the input. Defaults to an empty string.
6. `bordered` *(optional) boolean*
    * The input field border style. Defaults to true.
7. `disabled` *(optional) boolean*
    * Disables the button via a class on its wrapper, and an attribute on the input. Defaults to false.
8. `required` *(optional) boolean*
    * Adds/removes a supporting element, `<span>OPTIONAL</span>` to show the field is optional. Defaults to false.
9. `invalid` *(optional) boolean*
    * if true adds styles to denote if a field is invalid.
10. `prefillValue` *(optional) string*
    * Adds prefill data to the input field and adds custom styling for prefilled inputs. Defaults to an empty string.
11. `prefixContent` *(optional) string*
    * Adds a prefix to the input - can be supplied with another component (eg for an svg) which will be rendered. Defaults to false.
12. `suffixContent` *(optional) string*
    * Adds a suffix to the input - can be supplied with another component (eg for an svg) which will be rendered. Defaults to false.
13. `listId` *(optional) string* //TO DO
    * Exclusively for the Combo component datalist functionality. Links a datalist to the input field.
14. `autocomplete` *(optional) string*
    * Turn the browsers implementation of autocompletion/memory of forms on or off. Defaults to 'off'.
15. `handleChange` *(required) func*
    * Custom handler to attach to the input field - used to get the value of the field for example.

