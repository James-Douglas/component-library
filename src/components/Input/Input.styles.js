import React from 'react';
import styled, { css } from 'styled-components';
import MaskedInput from 'react-text-mask';

export const StyledWrapper = styled.div`
  margin-bottom: 2rem;
`;

export const StyledClearIcon = styled.button`
  position: absolute;
  top: 0;
  right: 0;
  width: ${({ theme }) => theme.spacing[44]};
  height: ${({ theme }) => theme.spacing[44]};
  transition: .2s ease-in-out all;
  pointer-events: ${({ disabled }) => (disabled ? 'none' : 'auto')};
  color: ${({ theme, isAutofill }) => (isAutofill ? theme.input.clearButton.colorAutofill : theme.input.clearButton.color)};
  :hover {
    color: ${({ theme }) => theme.input.clearButton.hoverColor};
  }
`;

export const StyledAffix = styled.span`
  font-weight: ${({ theme }) => theme.fontWeight.bold};
  font-size: ${({ theme }) => theme.fontSize.base};
  display: flex;
  align-items: center;
  justify-content: center;
  min-width:  ${({ theme }) => theme.spacing[48]};
  padding: ${({ theme }) => `${theme.spacing[8]} ${theme.spacing[12]}`};
  height: ${({ theme }) => theme.input.height};
  ${({ prefixBlock, isAutofill, theme }) => prefixBlock && !isAutofill && css`
    background: ${theme.colors.grey100};
  `}
`;

export const StyledInputContainer = styled.div`
  position: relative;
  width: 100%;
`;

export const StyledInputWrap = styled.div`
  display: flex;
  background: ${({ theme }) => theme.input.background};
  ${({ theme, isAutofill, disabled }) => (isAutofill && !disabled) && css`
    background: ${theme.colors.inputPrefilled};
  `}
  border: ${({ theme }) => theme.borders.transparent};
  :hover {
    border: ${({ theme, disabled }) => (disabled ? '' : theme.borders.hover)};
  }
  ::placeholder {
    ${({ theme }) => ({ ...theme.placeholder })};
  }
  [disabled] {
    background: ${({ theme }) => theme.input.background};
  }
  ${({ theme, bordered }) => bordered && css`
    border: ${theme.borders.component};
  `}
  ${({
    theme, bordered, isAutofill, disabled,
  }) => (bordered && isAutofill && !disabled) && css`
    border: ${theme.borders.prefill};
  `}
 
  ${({ disabled }) => (disabled) && css`
    opacity: 0.5;
  `}
  ${({ theme, isFocusActive }) => isFocusActive && css`
    border: ${theme.borders.hover};
  `}
   ${({ theme, invalid }) => invalid && css`
    border: ${theme.borders.invalid};
    :hover {
      border: ${theme.borders.invalid};
    }
  `}
`;

export const StyledInputClearWrap = styled.div`
  position: relative;
  width: 100%;
  min-height: ${({ theme }) => `calc(${theme.input.height} + 0.1rem)`};
  margin-right: 0.1rem;
`;

// eslint-disable-next-line react/jsx-props-no-spreading,react/display-name,react/prop-types
export const StyledInput = styled(React.forwardRef(({ isAutofill, ...props }, ref) => <MaskedInput {...props} ref={ref} />))`
  padding-left: ${({ theme }) => theme.spacing[12]};
  padding-right: ${({ theme }) => theme.spacing[36]};
  display: block;
  width: 100%;
  font-size: ${({ theme }) => theme.fontSize.base};
  border: ${({ theme }) => theme.borders.transparent};
  height: ${({ theme }) => theme.input.height};
  -moz-appearance: textfield;
  ::-webkit-outer-spin-button,
  ::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  ::-ms-clear {
    display: none;
  }
  :focus,
  :active,
  :hover {
    outline: 0;
  }
  ${({ theme, isAutofill, disabled }) => isAutofill && !disabled && css`
    background: ${theme.colors.inputPrefilled};
  `}
`;
