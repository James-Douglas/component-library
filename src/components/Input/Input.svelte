<svelte:options tag='ctm-input' />
<Fieldset {label} {tooltip} {..._props.fieldsetOptions}>
  <div class:disabled={disabled} class:hidden={hidden} class='input-wrap-default'>
    <div
      class='fix-wrap'
      class:invalid={invalid}
      class:manor-prefilled-border={autofill && !disabled}
      bind:this={fixWrap}
      class:input-border={bordered}>
      {#if prefix}
        <span class='prefix' class:prefix-no-border={!bordered} class:manor-prefilled={autofill && !disabled}>{@html prefix}</span>
      {/if}

      <div class='input-close-wrap'>
        <input
          bind:this={inputField}
          on:input={setValue}
          on:click={inputFieldClicked}
          {id}
          {type}
          {placeholder}
          {disabled}
          {required}
          class='input-default'
          class:manor-prefilled={autofill && !disabled}
          {value}
          list={listId}
          {autocomplete}
        />


        {#if value.length}
          <button on:click={clearInput} id='input-close-button' tabindex='0' class:lighter={!autofill} class:darker={autofill}>
            <div class='sr-only'>Clears the {label} field.</div>
            <Icon name={'closeCircle'} size={'1.6'}></Icon>
          </button>
        {/if}
      </div>

      {#if suffix}
        <span class='suffix' class:suffix-no-border={!bordered}>{@html suffix}</span>
      {/if}
    </div>
   </div>

    <span slot="supporting-elements" class="w-full relative">
      {#if listId}
        <slot name='data-list'></slot>
      {/if}

      <div class="supporting-elements">

        <slot></slot>
        {#if !required}
          <div>
            <span class="manor-optional-indicator manor-subscript">Optional</span>
          </div>
        {/if}
      </div>
    </span>
</Fieldset>

<script>
  import { onMount, onDestroy, createEventDispatcher } from 'svelte';
  import Fieldset from '../Fieldset/Fieldset.svelte';
  import Row from '../Grid/Row/Row.svelte';
  import Column from '../Grid/Column/Column.svelte';
  import Label from '../Label/Label.svelte';
  import Tooltip from '../Tooltip/Tooltip.svelte';
  import Icon from '../Icon/Icon.svelte';
  import observeProps from '../observeProps';
  import { isDesktop } from '@comparethemarketau/manor-config/utils/breakpoint';

  export let id;
  export let type = 'text';
  export let label = '';
  export let placeholder = '';
  export let bordered = true;
  export let disabled = false;
  export let required = false;
  export let invalid = false;
  export let autofill = false;
  export let prefix = '';
  export let suffix = '';
  export let value = '';
  export let listId;
  export let autocomplete = 'off';
  export let hidden = false;
  export let tooltip = {};

  const dispatch = createEventDispatcher();

  let inputField;
  let fixWrap;
  let _props;
  let enableLabelTooltip = !isDesktop();
  let inputWrapperElement;

  $: fieldsetOptions = $$props.fieldsetOptions || {};

  $: {
    _props = observeProps(
      [
        'id',
        'type',
        'label',
        'placeholder',
        'bordered',
        'disabled',
        'required',
        'invalid',
        'autofill',
        'value',
        'suffix',
        'prefix',
        'listId',
        'list',
        'tooltip'
      ], $$props);

    if (inputField) {
      Object.entries(_props).forEach(([key, value]) => {
        if (value === undefined || value === null) {
          inputField.removeAttribute(key);
        } else {
          inputField.setAttribute(key, value);
        }
      });
    }
  }

  function inputFieldClicked(event) {
    dispatch('inputFieldClicked', event);
  }

  function setValue(event) {
    value = event.target.value;
    dispatch('setValue', value);
  }

  function clearInput() {
    value = '';
    dispatch('setValue', value);
  }

  function resizeHandler() {
    enableLabelTooltip = !isDesktop();
  }

  function focusStyle() {
    fixWrap.style.border = 'solid #1780F3 0.1rem';
  }

  function blurStyle() {
    fixWrap.style.border = '';
  }

  onMount(() => {
    window.addEventListener('resize', resizeHandler);
    inputField.addEventListener('focus', focusStyle);
    inputField.addEventListener('blur', blurStyle);
  });

  onDestroy(() => {
    window.removeEventListener('resize', resizeHandler);
    inputField.removeEventListener('focus', focusStyle);
    inputField.removeEventListener('blur', blurStyle);
  });
</script>

<style>
  .input-wrap-default {
    @apply relative w-256;
  }
  .input-default {
    @apply w-256  pl-12 block text-base border border-solid border-transparent;
    min-height: 4.4rem;
    padding-right: 3.6rem;
  }
  input:invalid {
    @apply shadow-none;
  }
  input::placeholder {
    @apply italic text-base text-grey;
  }
  .fix-wrap {
    @apply flex border border-solid border-transparent;
  }
  .fix-wrap:hover {
    @apply border border-solid border-light-blue;
  }
  .fix-wrap-focus {
    @apply border border-solid border-light-blue;
  }
  .input-border {
    @apply border border-solid border-grey-light;
  }
  .input-border-right {
    border-right-color: theme(colors.grey-light);
  }
  .input-border-left {
    border-left-color: theme(colors.grey-light);
  }
  input:focus, input:active, input:hover {
    @apply outline-none;
  }
  input[disabled] {
    @apply bg-white cursor-not-allowed;
  }
  .disabled {
    @apply opacity-50;
  }
  .lighter {
    fill: theme(colors.grey-light);
  }
  .darker {
    fill: theme(colors.grey-dark);
  }
  #input-close-button:hover {
    fill: theme(colors.light-blue);
  }
  .input-default, .input-wrap-default {
   @apply w-full;
  }
  .input-close-wrap {
    @apply relative w-full;
    min-height: 4.4rem;
  }
  #input-close-button {
    @apply absolute w-40 h-40;
    right: 0.2rem;
    top: 0.2rem;
    transition: .2s ease-in-out all;
  }
  .supporting-elements {
    @apply flex justify-end w-full pt-8;
  }
  .manor-optional-indicator {
    margin-left: 1rem;
  }
  .prefix, .suffix {
    @apply flex items-center justify-center text-base font-bold;
    padding: 0.9rem 1.4rem;
    min-height: 4.4rem;
  }
  .prefix {
    padding-right: 0.5rem;
  }
  .suffix {
    padding-left: 0.5rem;
  }
  .invalid {
    @apply border border-solid border-invalid;
  }
  .prefix-no-border {
    @apply bg-white border border-solid border-transparent;
  }
  .suffix-no-border {
    @apply bg-white border border-solid border-transparent;
  }
  /* **********************************************************************
    Override Autocomplete styles in Chrome
    and include functionality for displaying pre-filled fields
  ************************************************************************* */
  .manor-prefilled-border {
    @apply border border-solid border-prechecked-darker;
  }
  .manor-prefilled,
  .input-default:-webkit-autofill,
  .input-default:-webkit-autofill:hover,
  .input-default:-webkit-autofill:focus {
    @apply bg-prechecked;
    -webkit-text-fill-color: theme(colors.black);
  }
  ::-ms-clear {
    display: none;
  }
</style>
