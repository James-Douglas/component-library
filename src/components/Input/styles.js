import css from 'styled-jsx/css';

export default css`
.input-container {
  @apply relative;
}

.input-default, .input-container {
  @apply w-full;
 }

.input-default {
  @apply pl-12 block text-base border border-solid border-transparent;
  border-width: 1px; /* IE 11 specific fix */
  height: 4.4rem;
  padding-right: 3.6rem;
}

input:invalid {
  @apply shadow-none;
}

input::placeholder {
  @apply italic text-base text-grey;
}

.input-wrap {
  @apply flex border border-solid border-transparent;
  border-width: 1px; /* IE 11 specific fix */
}

.input-wrap:hover:not(.disabled) {
  @apply border border-solid border-light-blue;
  border-width: 1px; /* IE 11 specific fix */
}

.input-border {
  @apply border border-solid border-grey-light;
  border-width: 1px; /* IE 11 specific fix */
}

.input-border-right {
  border-right-color: theme(colors.grey-light);
  border-width: 1px; /* IE 11 specific fix */
}

.input-border-left {
  border-left-color: theme(colors.grey-light);
  border-width: 1px; /* IE 11 specific fix */
}

input:focus, input:active, input:hover {
  @apply outline-none;
}

input[disabled] {
  @apply bg-white cursor-not-allowed;
}

.lighter {
  fill: theme(colors.grey-light);
}

.darker {
  fill: theme(colors.grey-dark);
}

.input-clear-wrap {
  @apply relative w-full;
  min-height: 4.4rem;
}

.input-clear-button {
  @apply absolute w-40 h-40;
  right: 0.2rem;
  top: 0.2rem;
  transition: .2s ease-in-out all;
}

.input-clear-button:hover {
  fill: theme(colors.light-blue);
}

.supporting-elements {
  @apply flex justify-end w-full pt-8;
}

.manor-optional-indicator {
  margin-left: 1rem;
}

.prefix, .suffix {
  @apply flex items-center justify-center text-base font-bold;
  padding: 0.9rem 1.4rem;
  height: 4.4rem;
}

.prefix {
  padding-right: 0.5rem;
}

.suffix {
  padding-left: 0.5rem;
}

.invalid {
  @apply border border-solid border-invalid;
  border-width: 1px; /* IE 11 specific fix */
}

.prefix-no-border {
  @apply bg-white border border-solid border-transparent;
}

.suffix-no-border {
  @apply bg-white border border-solid border-transparent;
}
.manor-prefilled {
  @apply bg-prechecked;
}
.manor-prefilled-border {
  @apply border border-solid border-prechecked-darker;
  border-width: 1px; /* IE 11 specific fix */
}

.input-wrap-focus {
  @apply border border-solid border-light-blue;
  border-width: 1px; /* IE 11 specific fix */
}

.disabled {
  @apply opacity-50;
}

.input-default:-webkit-autofill,
.input-default:-webkit-autofill:hover,
.input-default:-webkit-autofill:focus {
  -webkit-text-fill-color: theme(colors.black);
}

::-ms-clear {
  display: none;
}

.list {
  margin-top: -8px;
}
.list.up-list {
  margin-top: -22px;
}
`;
