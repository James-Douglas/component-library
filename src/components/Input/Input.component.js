import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import useValueState from '../../hooks/useValueState';
import Icon from '../Icon/Icon.component';
import Fieldset from '../Fieldset/Fieldset.component';
import styles from './styles.js';

/* Input will need to accept children for the custom combo */

export const renderClearIcon = (value, clearInput, isAutofill, label) => {
  if (value.length) {
    return (
      <>
        <style jsx>{styles}</style>
        <button
          onClick={clearInput}
          className={`
          input-clear-button
          ${isAutofill ? 'darker' : 'lighter'}
        `}
        >
          <div className="sr-only">Clears the {label}{' '}field.</div>
          <Icon name="closeCircle" size={1.6} />
        </button>
      </>
    );
  }
  return null;
};

export const renderAffix = (affixType, affixContent, bordered, isAutofill, disabled) => {
  
  if ( affixType && affixContent ) {

    return (
      <>
        <style jsx>{styles}</style>
        <span className={`
          ${affixType === 'prefix' ? 'prefix' : 'suffix'}
          ${!bordered && affixType === 'prefix' ? 'prefix-no-border' : ''}
          ${!bordered && affixType === 'suffix' ? 'suffix-no-border' : ''}
          ${isAutofill && !disabled ? 'manor-prefilled' : ''}
        `}
        >
          {affixContent}
        </span>
      </>
    );
  }
  return null;
};

export const renderOptionalElement = (required) => {
  if (!required) {
    return (
      <>
        <span className="manor-subscript">Optional</span>
      </>
    );
  }
  return null;
};

const Input = ({
  id, type, placeholder, prefillValue, autofill, required, disabled, bordered, invalid, prefixContent, suffixContent, label, tooltip,
}) => {
  const [ value, setValue, clearValue ] = useValueState(prefillValue || '');
  const [ isAutofill, setIsAutofill ] = useState(autofill)
  const inputWrapElement = useRef(null);
  
  const clearInput = () => {
    clearValue();
    setIsAutofill(false);
  };

  const handleChange = (e) => {
    setIsAutofill(false);
    setValue(e);
  }

  const addFocus = () => {
    inputWrapElement.current.classList.add('input-wrap-focus');
  };

  const removeFocus = () => {
    inputWrapElement.current.classList.remove('input-wrap-focus');
  };

  return (
    <>
      <Fieldset label={label} tooltip={tooltip}>
        <style jsx>{styles}</style>
        <div className="input-container">
          <div
            ref={inputWrapElement}
            className={`
            input-wrap
            ${isAutofill && !disabled ? 'manor-prefilled-border' : ''}
            ${bordered ? 'input-border' : ''}
            ${invalid ? 'invalid' : ''}
            ${disabled ? 'disabled' : ''}
          `}
          >

            {renderAffix('prefix', prefixContent, bordered, isAutofill, disabled)}

            <div className="input-clear-wrap">
              <input
                id={id}
                name={id}
                type={type}
                placeholder={placeholder}
                disabled={disabled}
                value={value}
                onChange={handleChange}
                autoComplete="off"
                onFocus={addFocus}
                onBlur={removeFocus}
                className={`
                  input-default
                  ${isAutofill && !disabled ? 'manor-prefilled' : ''}
                `}
              />

              {renderClearIcon(value, clearInput, isAutofill, label)}

            </div>

            {renderAffix('suffix', suffixContent, bordered, isAutofill, disabled)}

          </div>
          <div className="supporting-elements">
            {renderOptionalElement(required)}
          </div>
        </div>

      </Fieldset>
    </>
  );
};

Input.propTypes = {
  id: PropTypes.string.isRequired,
  type: PropTypes.string,
  placeholder: PropTypes.string,
  autofill: PropTypes.bool,
  required: PropTypes.bool,
  disabled: PropTypes.bool,
  bordered: PropTypes.bool,
  invalid: PropTypes.bool,
  prefix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  suffix: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
  ]),
  label: PropTypes.string,
  tooltip: PropTypes.object,
};

Input.defaultProps = {
  type: 'text',
  autofill: false,
  required: true,
  disabled: false,
  bordered: true,
  invalid: false,
};

export default Input;
