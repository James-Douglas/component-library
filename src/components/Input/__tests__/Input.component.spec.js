import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Input, { renderClearIcon, renderAffix, renderOptionalElement } from '../Input.component';

describe('renderClearIcon()', () => {
  // eslint-disable-next-line react/prop-types
  const ClearIconContainer = ({ value, clearInput, isAutofill, label }) => (
    <>
      {renderClearIcon(value, clearInput, isAutofill, label)}
    </>
  );

  it('does not render an clearIcon if the value.length is over 0', () => {
    const { container } = render(<ClearIconContainer value="" />);
    
    expect(container).toBeEmpty();
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders a clearIcon when the value.length is > 0', () => {
    const { container } = render(<ClearIconContainer value="input text" label='test input'/>);

    const svg = container.querySelector('svg');

    expect(svg).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

});
/* affixType, affixContent, bordered, isAutofill, disabled */
describe('renderAffix()', () => {
  // eslint-disable-next-line react/prop-types
  const AffixContainer = ({ affixType, affixContent, bordered, isAutofill, disabled }) => (
    <>
      {renderAffix(affixType, affixContent, bordered, isAutofill, disabled)}
    </>
  );

  it('does not render a prefix or suffix if its not supplied', () => {
    const { container } = render(<AffixContainer />);
    
    expect(container).toBeEmpty();
    expect(container.innerHTML).toMatchSnapshot();
  });

  
  it('renders a prefix when supplied', () => {
    const { container } = render(<AffixContainer affixType={'prefix'} affixContent={'?'}/>);
    
    const prefix = container.querySelector('.prefix');
    const suffix = container.querySelector('.suffix');

    expect(suffix).toBe(null);
    expect(prefix).toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders a suffix when supplied', () => {
    const { container } = render(<AffixContainer affixType={'suffix'} affixContent={'?'}/>);
    
    const prefix = container.querySelector('.prefix');
    const suffix = container.querySelector('.suffix');

    expect(prefix).toBe(null);
    expect(suffix).toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders a suffix when supplied with additional styling for autofill and border', () => {
    const { container } = render(<AffixContainer affixType={'prefix'} affixContent={'?'} border={false} isAutofill={true} disabled={false}/>);
    
    const prefix = container.querySelector('.prefix');
    const suffix = container.querySelector('.suffix');
    const border = container.querySelector('.prefix-no-border');
    const prefill = container.querySelector('.manor-prefilled');

    expect(suffix).toBe(null);
    expect(prefix).toBeInTheDocument();
    expect(border).toBeInTheDocument();
    expect(prefill).toBeInTheDocument();
    expect(container.innerHTML).toMatchSnapshot();
  });

});

xdescribe('renderOptionalElement()', () => {

});

/* 
  id, type, placeholder, prefillValue, autofill, 
  required, disabled, bordered, invalid, prefix, 
  suffix, label, tooltip, 
*/
xdescribe('Input.component', () => {
  it('renders with minimal props', () => {
    const { container } = render(<Input id="test-id" />);
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders with props', () => {
    const mockTestClick = jest.fn();
    const { container } = render(
      <Input
        id="test-id"
        icon="check"
        disabled={false}
        invertColour
        handleChange={mockTestClick}
      >
        <p>child content</p>
      </Input>,
    );

    const checkbox = container.querySelector('#test-id');
    const label = container.querySelector('label');
    const labelStyle = container.querySelector('.manor-checkbox');

    fireEvent.click(label, { button: 0 });

    expect(mockTestClick).toHaveBeenCalled();
    expect(labelStyle.classList).toContain('inverted');
    expect(checkbox.getAttribute('disabled')).toBe(null);
    expect(container.querySelector('#child-content')).toBeDefined();
    expect(container.getElementsByTagName('svg')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('checks on click', () => {
    const { container } = render(<Input id="test-id" icon="check" disabled={false} invertColour />);

    const checkbox = container.querySelector('#test-id');
    const label = container.querySelector('label');

    fireEvent.click(label, { button: 0 });

    expect(checkbox.checked).toBe(true);
  });

  it('click is disabled when disabled prop is supplied', () => {
    const { container } = render(<Input id="test-id" icon="check" disabled />);

    const checkbox = container.querySelector('#test-id');
    const label = container.querySelector('label');

    fireEvent.click(label, { button: 0 });

    expect(checkbox.checked).toBe(false);
  });
});
