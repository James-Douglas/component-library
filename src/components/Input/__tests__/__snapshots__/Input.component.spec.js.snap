// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderAffix() does not render a prefix or suffix if its not supplied 1`] = `""`;

exports[`renderAffix() renders a prefix when supplied 1`] = `
"<style></style><span class=\\"
          prefix
          prefix-no-border
          
          
        \\">?</span>"
`;

exports[`renderAffix() renders a suffix when supplied 1`] = `
"<style></style><span class=\\"
          suffix
          
          suffix-no-border
          
        \\">?</span>"
`;

exports[`renderAffix() renders a suffix when supplied with additional styling for autofill and border 1`] = `
"<style></style><span class=\\"
          prefix
          prefix-no-border
          
          manor-prefilled
        \\">?</span>"
`;

exports[`renderClearIcon() does not render an clearIcon if the value.length is over 0 1`] = `""`;

exports[`renderClearIcon() renders a clearIcon when the value.length is > 0 1`] = `
"<style></style><button class=\\"
          input-clear-button
          lighter
        \\"><div class=\\"sr-only\\">Clears the test input field.</div><svg version=\\"1.1\\" xmlns=\\"http://www.w3.org/2000/svg\\" x=\\"0px\\" y=\\"0px\\" enable-background=\\"0 0 44 44\\" class=\\" 
        icon 
         
         
      \\" viewBox=\\"0 0 44 44\\" style=\\"width: 1.6rem; height: 1.6rem;\\"><path d=\\"M22 .2C10 .2.2 10 .2 22S10 43.8 22 43.8 43.8 34 43.8 22 34 .2 22 .2zm9.9 29.4c.6.6.6 1.7 0 2.4-.6.6-1.7.6-2.4 0l-7.1-7.1-7.1 7.1c-.6.6-1.7.6-2.4 0-.6-.6-.6-1.7 0-2.4l7.1-7.1-7.1-7.1c-.6-.6-.6-1.7 0-2.4.6-.6 1.7-.6 2.4 0l7.1 7.1 7.1-7.1c.6-.6 1.7-.6 2.4 0 .6.6.6 1.7 0 2.4l-7.1 7.1 7.1 7.1z\\"></path></svg></button>"
`;
